### DISCUSSION TITLE
issue: useFieldArray remembering nested values after remove

### DISCUSSION THREAD

**Discussion by Wiklem** [react-hook-form/react-hook-form#8054] [Q&A] [answered]
### Version Number

7.28.0

### Codesandbox/Expo snack

https://codesandbox.io/s/objective-worker-ljigyt?file=/src/App.tsx

### Steps to reproduce

1. Go to https://codesandbox.io/s/objective-worker-ljigyt?file=/src/App.tsx
2. Click on Remove
3. Click on Append
4. The text should be cleared, but appear again after remove -> append.

### Expected behaviour

Expect that the object is empty or only has the given default values from append()

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
![Screen Shot 2022-03-18 at 9 26 16 pm](https://user-images.githubusercontent.com/10513364/158986403-3c991cdf-c0f0-424d-b39c-0e5fc20b8a33.png)

> Append input/inputs to the end of your fields and focus. The input value will be registered during this action.

> Important: append data is required and not partial.

**Comment by Wiklem** [comment]
> ![Screen Shot 2022-03-18 at 9 26 16 pm](https://user-images.githubusercontent.com/10513364/158986403-3c991cdf-c0f0-424d-b39c-0e5fc20b8a33.png)
> 
> > Append input/inputs to the end of your fields and focus. The input value will be registered during this action.
> 
> > Important: append data is required and not partial.

Thanks for reply.
So there is no way to actually remove the item?
As I write it and see in the controller code, the isDirty object hold the value even if its removed, and I have to "overwrite" it with append to make it disappear?
Its a bad solution in my case as the DTO is from backend and may change, so I end up setting values that should not be on the new appended object without even knowing it if I append the object with required data as stated.

**Comment by ldanet** [comment, marked as answer]
What I think Bill means is that the object you pass to the `append` function should specify all the keys as partial objects are not supported by the library. Since it's not supported, the behaviour is unspecified. In this instance it applies the default form values to the missing keys but even if that was what you wanted you shouldn't rely on this behaviour to be consistent.

For example you could do
```jsx
const emptyObject = {
  type: "AWAY",
  checked: false,
  goAway: null
}

[...]

<button type="button" onClick={() => append(emptyObject)}>
  Append
</button>
```
That way it's explicit what the values of a new object should be and it doesn't do weird unspecified behaviour.

https://codesandbox.io/s/amazing-booth-k2kkq3?file=/src/App.tsx

You don't always know what your backend object is, but you always know what your form is, don't you? Can you not do something like this?:
```jsx
<button type="button" onClick={() => append({ ...emptyObject, ...objectFromDTO })}>
  Append
</button>
```
That way you are sure that all the the fields you display in your form all have an appropriate default value that you have specified yourself even if the backend didn't, and if the backend did specify those values then it overrides your default ones. 

Edit: After re-reading your message I understand it's the other way around that bothers you, you don't necessarily want all the fields to be defined when you send them back. I feel like this maybe means that you are probably stretching the concept of a form and displaying values from an object that are not strictly form values, as in, they don't relate to a form input?

**Reply by T41WO** [reply]
Hi, I have the exact same experience and tried this solution. Passing a full default object to the append function. 

My form is one that loads preexisting list of values (for example: item[0], item[1], item[2]), passed in as defaultValues. If I remove at index 1, and then I append again, what I get is former index 2 in its place (item[0], item[2], item[2]). Subsequent append adds empty values instead of my full default object passed into append().

**Reply by T41WO** [reply]
I should mention that I am using v6.15.2. In my code sample, I switched the version to the latest and find that this solution worked. Is there any fixed for this using in v6.15.2? Not scheduled to migrate to v7 anytime soon.

code sample: https://codesandbox.io/s/0pnz0i?file=/src/App.tsx

**Comment by bluebill1049** [comment]
thanks, @ldanet that's correct. React hook form's useFieldArray is uncontrolled by default which means each newly appended input will need a default value.
